### PR TITLE
Option::flatten

### PR DESCRIPTION
**Pull request by eopb**
This PR makes this possible.
```rust
assert_eq!(Some(6), Some(Some(6)).flatten());
assert_eq!(Some(6), Some(Some(6)).into());
```

### CODE REVIEW

**Comment by Centril** [comment]
I think it would be preferable to add this as a method `.flatten()` since that is clearer wrt. intent and is better in terms of type inference. Moreover, we already have the other monadic operations and `.flatten()` already exists for `Iterator` so there's precedent there.

-------------

Assigning over to random t-libs member;

r? @alexcrichton

**Comment by ethanboxx** [comment]
@Centril I agree, a flatten method would be a good idea. Do you think I should add both a flatten method and a `Into` implementation?

**Comment by Centril** [comment]
> and an `Into` implementation?

Don't really have an opinion :)

**Comment by Keruspe** [comment]
You probably could use unwrap_or_default instead of matching, couldn't you?

**Comment by Centril** [comment]
> You probably could use unwrap_or_default instead of matching, couldn't you?

That imposes a constraint `T: Default`.

What you can do instead is `self.and_then(core::convert::identity)`,
since it holds that `(>>= id) = join`.

**Comment by Centril** [comment]
> 
> [00:04:14] error[E0433]: failed to resolve: use of undeclared type or module `core`
> [00:04:14]     --> src/libcore/option.rs:1435:23
> [00:04:14]      |
> [00:04:14] 1435 |         self.and_then(core::convert::identity)

You are inside libcore, so you have to use `crate::convert::identity`.

**Comment by Centril** [comment]
@ethanboxx once you're done with the changes, can you please squash them into a single commit?

**Comment by ethanboxx** [comment]
@Centril I dont understand this error. Can you explain why this is failing?

**Comment by Centril** [comment]
@ethanboxx It seems to me you are getting a type inference error. Try removing the implementation and that should fix it.

**Comment by ethanboxx** [comment]
@Centril Whats the best way to solve this new issue.

**Comment by cogmover** [comment]
TL;DR: Add `#![feature(option_flattening)]` at the start of the doc test, and move the attributes so they'll be below the doc comment (right above the method).
The first change fixes the issue, while the second one is to conform to the established style.

Documentation tests are external to the crate they're testing, so the `x.flatten()` in your doc test works the same way as writing `x.flatten()` in your own crate.
However, `Option::flatten` is unstable, so you'd have to opt into it with the feature gate, so adding `#[feature(option_flattening)]` at the start is what you need to solve the issue.
`Vec` has some unstable methods as well, and you can use them as a reference to make sure you're doing everything right.

In other places in the standard library, the `#[inline]` and `#[unstable]` attributes are below the documentation comment, just above the method itself. @Centril, is the "doc comments above attributes" style official and documented somewhere? I'm asking you because you've commented here before (and are generally helpful :smile:), but if you're not the right person to ask just redirect the question to whoever is.

**Comment by ethanboxx** [comment]
@Centril @cogmover @Keruspe Thanks for all the invaluable help. Looks like everything is working now.

**Comment by alexcrichton** [comment]
Thanks for the PR @ethanboxx! I'm personally a bit wary on adding this method though since I feel like what it morally is striving to do is collapse any amount of inner options, but it doesn't handle, for example, `Option<Option<Option<T>>> -> T`. This definitely works for two layers deep, but I could see how ergonomically it's intended (and the name sort of implies) that it folds everything on the inside.

The `Option` type is also very core to Rust and additions to its API shouldn't be taken lightly, since now Rust programmers will have to recognize what `.flatten()` is doing when they see it in the wild (another method to keep in mind with `Option`). Now that of course doesn't mean we can't extend its API, just that we need to be careful in doing so!

**Comment by Centril** [comment]
@alexcrichton 

> but it doesn't handle, for example, `Option<Option<Option<T>>> -> T`. This definitely works for two layers deep, but I could see how ergonomically it's intended (and the name sort of implies) that it folds everything on the inside.

Note that this is the same for `iter.flatten()`. That's just how monadic join works. Is there a particular difference wrt. `Option`? Note that unnesting was also discussed in the tracking issue for `Iterator::flatten`: https://github.com/rust-lang/rust/issues/48213.

**Comment by ethanboxx** [comment]
@alexcrichton You are correct. My original intention was for it flatten all nested layers, however, it quickly became apparent that it was not possible.  I have added extra documentation to clarify this in a similar way `iter.flatten()` is documented.

**Comment by alexcrichton** [comment]
Ok, seems reasonable to me!

@bors: r+
